packages/router/__tests__/blocking-test.ts

```diff
@@ -0,0 +1,322 @@
+import type { Router } from "../index";
+import { createMemoryHistory, createRouter } from "../index";
+
+describe("blocking", () => {
+  let router: Router;
+  it("creates a blocker", () => {
+    router = createRouter({
+      history: createMemoryHistory({
+        initialEntries: ["/"],
+        initialIndex: 0,
+      }),
+      routes: [{ path: "/" }, { path: "/about" }],
+    });
+    router.initialize();
+
+    let fn = () => true;
+    router.createBlocker("KEY", fn);
+    let blocker = router.getBlocker("KEY");
+    expect(blocker).toEqual({
+      fn,
+      state: "unblocked",
+      proceed: undefined,
+      reset: undefined,
+    });
+  });
+
+  it("deletes a blocker", () => {
+    router = createRouter({
+      history: createMemoryHistory({
+        initialEntries: ["/"],
+        initialIndex: 0,
+      }),
+      routes: [{ path: "/" }, { path: "/about" }],
+    });
+    router.initialize();
+
+    router.createBlocker("KEY", () => true);
+    router.deleteBlocker("KEY");
+    let blocker = router.getBlocker("KEY");
+    expect(blocker).toBeUndefined();
+  });
+
+  describe("on history push", () => {
+    let initialEntries = ["/", "/about"];
+    let initialIndex = 0;
+    beforeEach(() => {
+      router = createRouter({
+        history: createMemoryHistory({
+          initialEntries,
+          initialIndex,
+        }),
+        routes: [{ path: "/" }, { path: "/about" }],
+      });
+      router.initialize();
+    });
+
+    describe("blocker returns false", () => {
+      it("sets blocker state to 'unblocked'", async () => {
+        let blocker = router.createBlocker("KEY", () => false);
+        await router.navigate("/about");
+        expect(blocker.state).toEqual("unblocked");
+      });
+
+      it("navigates", async () => {
+        router.createBlocker("KEY", () => false);
+        await router.navigate("/about");
+        expect(router.state.location.pathname).toBe("/about");
+      });
+    });
+
+    describe("blocker returns true", () => {
+      it("set blocker state to 'blocked'", async () => {
+        let blocker = router.createBlocker("KEY", () => true);
+        await router.navigate("/about");
+        expect(blocker.state).toEqual("blocked");
+      });
+
+      it("does not navigate", async () => {
+        router.createBlocker("KEY", () => true);
+        await router.navigate("/about");
+        expect(router.state.location.pathname).toBe(
+          initialEntries[initialIndex]
+        );
+      });
+    });
+
+    describe("proceeds from blocked state", () => {
+      it("sets blocker state to 'proceeding'", async () => {
+        let blocker = router.createBlocker("KEY", () => true);
+        await router.navigate("/about");
+        expect(blocker.state).toEqual("blocked");
+
+        await blocker.proceed?.();
+        expect(blocker.state).toEqual("proceeding");
```

Comment:
Do we want a `blocker.state === 'unblocked'` assertion once it completes the navigation?